scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5217 @@
+        # depend on others. If the relationshp is simple (e.g. a < b < c),
+        # we can draw values in order a, b, c.
+        parameter_values = {}
+
+        if not len(sizes) or not np.iterable(sizes[0]):
+            sizes = [sizes]*len(self.parameters)
+
+        for size, param in zip(sizes, self.parameters.values()):
+            parameter_values[param.name] = param.draw(
+                size, rng=rng, proportions=proportions,
+                parameter_values=parameter_values)
+
+        return parameter_values
+
+
+def _set_invalid_nan(f):
+    # Wrapper for input / output validation and standardization of distribution
+    # functions that accept either the quantile or percentile as an argument:
+    # logpdf, pdf
+    # logcdf, cdf
+    # logccdf, ccdf
+    # ilogcdf, icdf
+    # ilogccdf, iccdf
+    # Arguments that are outside the required range are replaced by NaN before
+    # passing them into the underlying function. The corresponding outputs
+    # are replaced by the appropriate value before being returned to the user.
+    # For example, when the argument of `cdf` exceeds the right end of the
+    # distribution's support, the wrapper replaces the argument with NaN,
+    # ignores the output of the underlying function, and returns 1.0. It also
+    # ensures that output is of the appropriate shape and dtype.
+
+    endpoints = {'icdf': (0, 1), 'iccdf': (0, 1),
+                 'ilogcdf': (-np.inf, 0), 'ilogccdf': (-np.inf, 0)}
+    replacements = {'logpdf': (-oo, -oo), 'pdf': (0, 0),
+                    '_logcdf1': (-oo, 0), '_logccdf1': (0, -oo),
+                    '_cdf1': (0, 1), '_ccdf1': (1, 0)}
+    replace_strict = {'pdf', 'logpdf'}
+    replace_exact = {'icdf', 'iccdf', 'ilogcdf', 'ilogccdf'}
+    clip = {'_cdf1', '_ccdf1'}
+    clip_log = {'_logcdf1', '_logccdf1'}
+
+    @functools.wraps(f)
+    def filtered(self, x, *args, **kwargs):
+        if self.iv_policy == _SKIP_ALL:
+            return f(self, x, *args, **kwargs)
+
+        method_name = f.__name__
+        x = np.asarray(x)
+        dtype = self._dtype
+        shape = self._shape
+
+        # Ensure that argument is at least as precise as distribution
+        # parameters, which are already at least floats. This will avoid issues
+        # with raising integers to negative integer powers and failure to replace
+        # invalid integers with NaNs.
+        if x.dtype != dtype:
+            dtype = np.result_type(x.dtype, dtype)
+            x = np.asarray(x, dtype=dtype)
+
+        # Broadcasting is slow. Do it only if necessary.
+        if not x.shape == shape:
+            try:
+                shape = np.broadcast_shapes(x.shape, shape)
+                x = np.broadcast_to(x, shape)
+                # Should we broadcast the distribution parameters to match shape of x?
+                # Should we copy if we broadcast to avoid passing a view to developer
+                # functions?
```

Comment:
Thanks for the context. I'd suggest to remove the comment before merging, because for now it doesn't seem needed and it raises more questions than it answers.

Reply:
Removed in next commit. Broadcasted views are typically read-only anyway.